File: App/Betterboxd.hs
module App.Betterboxd where

import App.Models.Movie (Movie, idtM, tittle, rating, genres, year, actors, directors,comentarios, createMovie)
import App.Models.User ( User, createUser, idt, user, nome, bio, senha, listas )
import App.Util.StringsOp (concatStrings)
import App.Controllers.ListaController (createListaData, exibeListas, getLastLista)
import App.Controllers.MovieController (getMovies, getMoviesByGenre, getMoviesByTittle, appendComment, updteComment, editComment)
import App.Controllers.UserController (getUsers, appendUser, getUserLogged, getUserBy, hasUsername, getNextIdt, stringToUser, editUser, appendListaToUser)
import App.Data.CsvManager ( writeCSV )
import qualified Data.Maybe
import System.IO
import App.Models.Lista (Lista)

cadastraUsuario :: String -> String -> String -> String -> IO()
cadastraUsuario nome user bio senha = do
    createListaData ("Favoritos de " ++ nome)
    createListaData ("Filmes Loggados de " ++ nome)
    createListaData ("Recomendados para " ++ nome)
    appendUser [user, nome, bio, senha]

isLoginValid :: String -> String -> Bool
isLoginValid usern psswrd = Data.Maybe.isJust usuarioInformado && (senha (Data.Maybe.fromJust usuarioInformado) == psswrd)
    where usuarioInformado = getUserBy (getUsers 0) user usern

doLogin :: String -> IO()
doLogin userName = do
    let usuario = Data.Maybe.fromJust $ getUserBy (getUsers 0) user userName
    writeCSV "./App/Data/Temp.csv" [[idt usuario]]

commentMovie :: User -> Int -> String -> Movie -> IO()
commentMovie usr nStar comment mvie = do
    appendComment (idt usr) nStar comment mvie

changeComment :: Movie -> (String, Int, String) -> IO()
changeComment mvie (a, b, c) = updteComment novosComment mvie
    where novosComment = editComment (comentarios mvie) (a,b,c)

printMovieInfo :: Movie -> IO ()
printMovieInfo movie = do
    putStrLn "\ESC[2J"
    putStrLn $ replicate 80 '*'
    putStrLn ""
    putStrLn $ replicate 30 ' ' ++ "Informações do Filme"
    putStrLn ""
    putStrLn $ replicate 80 '*'
    putStrLn $ " Título:          " ++ tittle movie
    putStrLn $ " Rating:          " ++ show (rating movie)
    putStrLn $ avaliacaoBbxd (comentarios movie)
    putStrLn $ " Ano:             " ++ show (year movie)
    putStrLn $ " Atores:          " ++ concatStrings (actors movie) ", "
    putStrLn $ " Diretor:         " ++ concatStrings (directors movie) ", "
    putStrLn $ replicate 80 '*'
    putStrLn " Comentários:"
    if null (comentarios movie)
        then putStrLn "  - Nenhum comentário disponível."
        else mapM_ (\(usuario, avaliacao, comentario) ->
                putStrLn $ " " ++ nome (Data.Maybe.fromJust (getUserBy (getUsers 0) idt usuario)) ++ " - " ++ stars avaliacao ++ " " ++ comentario)
              (comentarios movie)
    putStrLn $ replicate 80 '*'
    where
        stars :: Int -> String
        stars n = replicate n '*' ++ replicate (5 - n) '°'

        avaliacaoBbxd :: [(String, Int, String)] -> String
        avaliacaoBbxd [] = " Avaliação média: Nenhuma avalição realizada"
        avaliacaoBbxd tuplas =
            let valores = map (\(_, x, _) -> x) tuplas
                soma = sum valores
                total = fromIntegral (length valores)
                media = div soma  total
            in " Avaliação média: " ++ show media ++ " estrelas"



verificaComentUnico :: String -> [(String, Int, String)] -> Bool
verificaComentUnico _ []            = True
verificaComentUnico str [(x, y, z)] = str /= x
verificaComentUnico str ((x, y, z):xs)
    | str == x      = False
    | otherwise     = verificaComentUnico str xs

criaLista :: String -> User -> IO()
criaLista str usr = do
    createListaData str
    appendListaToUser (getLastLista 0) usr

File: Prolog/App/betterboxd.pl
:- use_module(library(csv)).
:- encoding(utf8).
:- set_prolog_flag(encoding, utf8).
:- consult('App/Controllers/ListaController.pl').
:- consult('App/Controllers/UserController.pl').
:- consult('App/Controllers/MovieController.pl').


cadastraUsuario(Nome, User, Bio, Senha):-
    Temp1 = ['Favoritos de ', Nome],
    Temp2 = ['Filmes loggados de ', Nome],
    Temp3 = ['Recomendados para ', Nome],
    atomic_list_concat(Temp1, '', Nome1),
    atomic_list_concat(Temp2, '', Nome2),
    atomic_list_concat(Temp3, '', Nome3),
    createLista(Nome1, T1),
    createLista(Nome2, T2),
    createLista(Nome3, T3),
    atomic_list_concat([T1, T2, T3], ',', IdsLista),
    appendUser(User, Nome, Bio, Senha, IdsLista).

isLoginValid(User, Senha):-
    hasUsername(User, row(_, User, _, _, Senha, _)).

doLogin(Username):-
    hasUsername(Username, row(Id, Username, _, _, _, _)),
    csv_write_file('App/Data/Temp.csv', [row(Id)]).

criaLista(row(Id, Username, Name, Bio, Senha, IdsLista), Nome, NovoUser):-
    createLista(Nome, T),
    Temp = [IdsLista, T],
    atomic_list_concat(Temp, ',', Resposta),
    editUser(Id, Username, Name, Bio, Senha, Resposta),
    getUser(Id, NovoUser).

media_avaliacoes([], 0, 0).
media_avaliacoes([row(_, _, NumStar, _) | Resto], Soma, NumAvaliacoes) :-
    media_avaliacoes(Resto, SomaResto, NumAvaliacoesResto),
    Soma is SomaResto + NumStar,
    NumAvaliacoes is NumAvaliacoesResto + 1.

media_total_avaliacoes(Lista, Media) :-
    media_avaliacoes(Lista, Soma, NumAvaliacoes),
    NumAvaliacoes > 0,
    Media is Soma / NumAvaliacoes.

avaliacoes([],  ' Avaliação média: Nenhuma avalição realizada'):- !.
avaliacoes(Lista, Str):-
    media_total_avaliacoes(Lista, Media),
    Temp = [' Avaliação média: ', Media, ' estrelas'],
    atomic_list_concat(Temp, Str).

comentarios([], '  - Nenhum comentário disponível.'):- !.
comentarios(P, Str):-
    comentariosAux(P, Lista),
    atomic_list_concat(Lista, Str).

comentariosAux([], []).
comentariosAux([row(IdUser, _, NumStar, Comentario)|T], L):-
    getUser(IdUser, row(IdUser, _, Name, _, _, _)),
    comentariosAux(T, L2),
    Temp = [' ', Name, " - ", NumStar, ' estrelas: ', Comentario, '\n'],
    append(Temp, L2, L).
    
movieInfo(row(Id, Titulo, Rating, _, Ano, Atores, Diretores, _), Str):-
    getComentariosByMovie(Id, Comentarios),
    avaliacoes(Comentarios, Temp1),
    comentarios(Comentarios, Temp2),
    Lista = [
                '********************************************************************************\n',
                '\n',
                '                              Informações do Filme\n',
                '\n',
                '********************************************************************************\n',
                ' Título:          ', Titulo, '\n',
                ' Rating:          ', Rating, '\n',
                Temp1, '\n',
                ' Ano:             ', Ano, '\n',
                ' Atores:          ', Atores, '\n',
                ' Diretor:         ', Diretores, '\n',
                '********************************************************************************\n',
                ' Comentários:\n',
                Temp2, '\n',
                '********************************************************************************\n'
    ],
    atomic_list_concat(Lista, Str).


